Add Deprecation Process DIP

JackStouffer opened the pull request:
cc @wilzbach @andralex @CyberShadow

CyberShadow commented:
Thanks!

Recent debate on using dates vs. versions for deprecation deadlines:
https://github.com/dlang/phobos/pull/6086#discussion_r164306440

CC @jmdavis, the deprecations czar.

wilzbach commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,75 @@
+Users must be given at least four major releases before the deprecated symbols
+are removed. More releases should be given if the removed code is commonly used.
+
+On the third release, the documentation for the symbol should be removed while
```
We have the https://docarchives.dlang.io which are index by Google too, so we could do this earlier.

wilzbach commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,75 @@
+specific release. This allows users to easily know if upgrading will break their
+code or not.
+
+Both at the time of deprecation and removal, a changelog entry must be made.
```
Maybe add something like: 

> This changelog entry should have a short motivation for the deprecation (or removal) and needs to describe what actions are required by the user to upgrade their codebase.

wilzbach commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,75 @@
+in advance of yielding a deprecation message by default. If this approach is used,
+users must be given at least two major releases before not using the flag gives
+deprecation messages. The transition flag would then have no effect, and turn
+into an error when the deprecated feature is finally removed.
```
Imho the entire idea of `-transition` is nice, but completely failed in the past.
See e.g. the story of `-transition=complex`.
tl;dr: `-transition=complex` existed for 1.5 years and one couldn't even use standard library with it until recently https://github.com/dlang/dmd/pull/7640.

Hence, I suggest to drop the idea of `-transition` for future deprecation and do the same as we do in Phobos and 
have recently done in dmd too(https://github.com/dlang/dmd/pull/7760/files#diff-90e876a104f01c31442b3b307953221cR1173): just trigger the deprecation.

wilzbach commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,75 @@
+deprecation and removal plan. The documentation should contain information to help
+the users using the symbol(s) transition their code away from the symbol(s).
+
+Users must be given at least four major releases before the deprecated symbols
```
We don't follow SemVer at the moment. With the closest approximation (`2.B.C`), bumping `B` is called a minor release :scream: 

However, @MartinNowak plans to change the release process soon. It would mean that the meaning of "release" changes from every two months to six months. So we need to be a bit careful here. OTOH four versions is a good value for now and if Martin's release process change (DIP?) goes through, this can be easily adjusted.
Though, I really like our current bi-monthly releases cycle, it's a good measure of stable and getting new stuff every now and then.

andralex commented:
cc @mdparker

jmdavis commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,86 @@
+are removed. More releases should be given if the removed code is commonly used.
+
+On the second release, the documentation for the symbol should be removed while
+keeping it public.
```
This is way too fast. For years now, the policy has been to have a symbol deprecated but documented for a year and then deprecated and undocumented for a year. That has gotten complaints both that it is too quick and that it is too long, but it's tried to balance that. It's long enough that even projects that are rarely maintained have a decent chance to be updated but not so long that deprecated stuff is around forever.

As it stands, four major releases is only going to be something like 8 - 10 months - maybe 12 if we run into several delayed releases. And one release of being deprecated before being removed from the docs means that anyone who is not completely on top of things is just going to find the documentation missing the next time they rebuild their project. Even if we all agreed that the current process is too slow, I think that this is clearly too fast - especially when you consider that plenty of our users aren't constantly updating their compilers. Heck, with the documentation only being there for one major release, it'll probably be gone before the ldc guys update to that release, since they're frequently about a major release behind.

This should probably be more like 5 major releases as deprecated and documented and 5 as deprecated and undocumented with the caveat that it can be adjusted in either direction if deemed appropriate by the development team but that it should only be done so with very good reason (e.g. if having the symbol around is causing problems, it should probably go sooner, but that's rarely the case).

mdparker replied:
That's a debate to have in the community review, methinks.

jmdavis commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,86 @@
+
+## Copyright & License
+
+Copyright (c) 2017 by the D Language Foundation
```
It's 2018 now.

jmdavis commented:
Also, if we're deprecating a symbol, because we're replacing it, if possible, the new symbol should be introduced at least one release prior to deprecating the old one so that it's possible to build your project without deprecation messages on both the latest release and master.

JackStouffer commented:
Addressed comments and added more details about language changes and exceptions to these rules.

jmdavis commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,109 @@
+## Rationale
+
+There is general disagreement on the best and/or accepted way to remove public
+features. Each deprecation ends being handled slightly differently depending on
```
ends up

wilzbach commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -66,21 +83,27 @@ code if refactoring is not possible.
+world systems, all changes or removals must be accompanied with a deprecation
+period. "Language features" includes bugs in the current behavior that existing
+user code depends on, e.g. [Issue 10378](https://issues.dlang.org/show_bug.cgi?id=10378).
+Fixing such issues should include a deprecation preiod for the current behavior,
```
Typo: Period

JackStouffer replied:
Do you mean this sentence should be broken up?

jmdavis replied:
No, you misspelled period. You wrote preiod.

JackStouffer replied:
Oh, ha. Totally missed that.

jmdavis replied:
I did too until Seb pointed it out.

jmdavis commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,109 @@
+There is general disagreement on the best and/or accepted way to remove public
+features. Each deprecation ends being handled slightly differently depending on
+who's handling the pull requests. Standardizing the process makes sure that
+deprecations are done very publicly and carefully, as to minimize breakage and
```
so as to

jmdavis commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,109 @@
+specific release. This allows users to easily know if upgrading will break their
+code or not.
+
+Users must be given at least eight non-patch releases before the deprecated features
```
ten deprecated with five undocumented would be better, since if we're actually on target for all releases, eight would only be 16 months rather than 24 like we do now. Ten would still only be 20 months if all releases are on time, but they get delayed frequently enough that that will probably push it into the 2 year range which is where we want to be. So, with 10 releases, we might be removing stuff a bit early, but with eight, it's pretty much a guarantee that we will unless releases start falling behind significantly.

wilzbach commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,109 @@
+both the `stable` and `master` branches.
+
+On the fourth release in the deprecation period, the documentation for the symbol
+should be removed while keeping the code itself public until removal.
```
How about adding a global summary table of the required actions?
Quick proposal:

action | when
---------| --------
add deprecation warning | immediately
add deprecation message to the docs | immediately
add deprecation changelog | immediately
remove deprecation from the docs | Fourth release
remove symbol + add final removal warnings | Fifth release

wilzbach commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,109 @@
+Deprecations to language features must also update the [language deprecations
+page](https://dlang.org/deprecate.html) on dlang.org simultaneously. The deprecation
+message given by the compiler should contain the planned removal period and/or a
+pointer to more information pertaining to the deprecation.
```
Walter doesn't like short urls, however I do like them and would love to use them. AFAICT we have never included a URL in a deprecation message. So we probably need to talk about this.

If short urls aren't possible, we need an alternative.
Idea: `dep.dlang.io/<alias>` (with e.g. redirect to a changelog entry)
I could setup `dep.dlang.io` with GH Pages, s.t. getting a new alias would be a matter of a simple one-line change PR.

JackStouffer replied:
When I mentioned "pointer", I had in mind something like "see the 2.099 changelog for details"

jmdavis replied:
Then you should probably give an example along those lines to make it clear.

JackStouffer commented:
Ping @mdparker

mdparker commented:
@JackStouffer I'm waiting on the new process document to be finalized before I move forward on any more DIPs. Waiting for feedback at the moment. Hopefully not much longer.

wilzbach commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,110 @@
+an image of instability.
+
+At the time of the pull request for deprecation, all code in Phobos, Druntime,
+and DMD must be updated to remove use of the effected code. Any projects that
```
FWIW while I successfully introduced `-de` to DMD, Phobos and Tools, DRuntime doesn't build with `-de` yet (https://github.com/dlang/druntime/pull/1982)
Definitely not a requirement for this DIP, just an FYI.

wilzbach commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,110 @@
+"See the 2.080 changelog for more details" or "See the function documentation for more
+details". The documentation of the symbol(s) must be updated noting the
+deprecation and removal plan. The documentation should contain information to help
+the users using the symbol(s) transition their code away from the symbol(s).
```
I think an example of a function signature with a deprecation message would be good here.

wilzbach commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,110 @@
+The symbol(s) must be marked using the `deprecated` keyword with a message containing
+the planned removal period. A pointer to more information should also be added. E.g.
+"See the 2.080 changelog for more details" or "See the function documentation for more
+details". The documentation of the symbol(s) must be updated noting the
```
Ouh so you were talking about this. As a user, this is a bit annoying.
OK, so I have to search for the changelog or function signature and hope I can find it. This takes at least three hops and is unnecessarily challenging.
I love the error messages from ShellCheck, e.g.

https://github.com/koalaman/shellcheck/wiki/SC2206

Can't we emit a similar deprecation message code? Phobos, dmd and druntime could use different leading numbers to avoid sync issues, but if numbers are too problematic because of merge conflicts (bumping the number), short keywords like DEPHEXSTRING or DEPENFORCEEX could work too?

JackStouffer replied:
>https://github.com/koalaman/shellcheck/wiki/SC2206

Maybe I'm missing something but I don't see anything about error messages on that page.

>short keywords like DEPHEXSTRING or DEPENFORCEEX could work too?

That's basically just an error code, which really isn't helpful either. At the end of the day, to get details on the deprecation/migration path you still would have to go to google and type that in. In the case described in the DIP, at least there's a specific place the user is told to go to.

jmdavis replied:
Normally, if you get a deprecation message, then you just look at the documentation for that symbol. I see no reason to complicate this with something like error codes (something that Walter has already expressed a dislike for BTW).

The deprecation message tells you that the symbol has been deprecated, and if the explanation is short enough, it also tells you what to use instead. If you want to know more, you look at the documentation for that symbol. For those who read the changelog, a changelog entry can give a heads up, but ultimately, it's the documentation on the symbol that's telling you what's going on. Let's please not complicate this.

mihails-strasuns-sociomantic replied:
For that to work better, it would be convenient to provide persistent link to documentation entry for the specific version used.

wilzbach replied:
> For that to work better, it would be convenient to provide persistent link to documentation entry for the specific version used.

Do you mean that the changelog entries should use links like e.g. this one?

https://docarchives.dlang.io/v2.079.0/phobos/std_functional.html#binaryReverseArgs

mihails-strasuns-sociomantic replied:
No, I mean actual deprecation messages printed by compiler. Less steps to take from message to relevant instructions -> more developers who actually read instructions as opposed to guessing -> less frustration

quickfur replied:
What @mihails-strasuns-sociomantic requets requires (1) dlang.org keep an archive (lasting at least as long as the oldest deprecations) of all documentation for each release, and (2) maintain a consistent URL scheme that doesn't change over time.  Both are good things, IMO, and we should definitely move in that direction.  Then we can put actual URLs in the deprecation messages without the chaos that will ensue if the contents of the URL changes over time or if the URL itself becomes invalid over time.  (Keep in mind that the URL must work even for previous releases of the compiler, so things like fixing a typo in the URL, etc., need to be done with care, e.g., with a webserver redirect in the right places.)

wilzbach commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,110 @@
+period.
+
+Deprecations to language features must also update the [language deprecations
+page](https://dlang.org/deprecate.html) on dlang.org simultaneously. The deprecation
```
I think we need to clarify how the docs need to be updated too, e.g. 

Deprecation: update the deprecatation list + add a warnings to the spec.

Error: remove from grammar + update deprecated list

wilzbach commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,110 @@
+2. The existance of the current code precludes its own fix or the fix of an equally
+important issue.
+
+Shorting the deprecation preiod should be done with caution to avoid giving D
```
s/preiod/period/

jmdavis commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,110 @@
+specific release. This allows users to easily know if upgrading will break their
+code or not.
+
+Users must be given at least 10 non-patch releases before the deprecated features
```
This is part of why we haven't been using releases for deprecations but have been using dates. Dates are consistent. Releases haven't been. The number 10 is based on the idea that we're getting a new major release approximately every other month but probably slip occasionally, so it's about 2 years. If we're going to change the release cycle, then this number isn't appropriate. And if we're not going to be at least reasonably consistent about when releases are, IMHO, it makes no sense whatsoever to use releases as the measurement for the deprecation cycle.

That being said, I hate the idea of major releases only coming out every 6 months. Not only would that suck from the standpoint of getting new stuff, but that effectively means that deprecations can only be introduced twice a year. And while ideally, we would only be deprecating stuff rarely, that could seriously delay us being able to get rid of something that we decide that we need to get rid of.

JackStouffer replied:
Releasing new versions has been a huge benefit to D IMO, and has done great things for other software like Firefox. I believe we move away from it at our detriment.

mihails-strasuns-sociomantic replied:
Well, I can say for certain that in our case it is extremely unlikely to upgrade _major_ compiler version more often than once in a year (as the process itself usually takes several months across different projects). With that in mind 2 years is a very useful term because it ensures that you can't possibly miss deprecation of some feature by skipping 1 year worth of versions.

Actual amount of major versions that cover that period does not matter that much in this context (less often major versions are beneficial for other reasons but it not within scope of this DIP).

JackStouffer commented:
Ping

mdparker commented:
@JackStouffer  Thanks for your patience. I'm moving on this stuff again. Once I get the older ones sorted, I'll get to this one. I ask that you first please update it to the layout in the new Template.md.

wilzbach commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,121 @@
+both the `stable` and `master` branches.
+
+On the fifth release in the deprecation period, the documentation for the symbol
+should be removed while keeping the code itself public until removal.
```
Does it need to be that long?
Once a symbol is removed from the documentation, the biggest cost is paid as it only triggers the deprecation messages, but no longer affects the learning process of newcomers (or other people looking at the doc for help).
If the argument is for the old symbols being searchable via Google, then there are now the docarchives.dlang.io with an indexed snapshot of every release + the changelog entries on dlang.org

Many projects also simply undocument their deprecated symbols with their removal (and sometimes put them on a special deprecated list) which imho is a better way - especially given that we constantly seem to forget to undocument the symbols - but I guess that's too much to ask for :/

jmdavis replied:
Yes, and much as I hate to say it, I'm honestly sick of discussing this. This is the length that we came to after discussion on quite a few occasions and complaints both about it being too short and too long. We need to balance the ability to remove symbols that we don't want around anymore and not breaking existing code. Some folks in the past have pushed for _never_ removing deprecated symbols, and both Andrei and Walter have expressed opinions along those lines at one time or another. For instance, at one point, Walter went to update one of his programs, and it didn't compile anymore, because some symbols had been deprecated and removed since the last time he'd updated it, and he was very unhappy about it. Those symbols had gone through the full two year deprecation process. Imagine how much more complaining we'd get along those lines if the length of the deprecation process were short. Every time that the length of the deprecation process comes up for discussion, I'm worried that Walter or Andrei is going to put their foot down and insist that it's not a problem to keep symbols around forever, and I really don't want to see that happen.

Various devs (usually those who are constantly up-to-date with the compiler that they use and probably who use master most of the time) like the idea of quickly deprecating stuff and then removing it, whereas other folks aren't as on top of things, and for them, fast deprecations and removals cause them problems. And the fact that ldc or gdc are not always being caught up with where dmd is just make that worse. Those needs need to be balanced, and if anything, over time, as our user base has grown and the amount of D code that's been written has grown, the need to avoid breaking code has grown.

After trial and error, we came to the process of having symbols deprecated and documented for one year and then deprecated and undocumented for one year, and over all, that's worked fairly well. It balances the needs of being able to remove old symbols while not breaking code, and it balances the need for folks to look at the documentation for the symbols they're replacing and the need to discourage the use of such symbols by not having them documented.

Yes, if there's a symbol that we really want to get rid of, the length of the deprecation process can be annoying, but we shouldn't be deprecating symbols very frequently, because that's just going to make folks complain about how D isn't stable, and it's a sign that we keep screwing up what we put in Phobos. But at least we have a process for removing such symbols, even if it's not lightning fast. There are other languages, that tend to be far more rigid with their standard library, and some stuff sits there _forever_ even if it's deprecated.

I won't claim by any means that our deprecation process is perfect, but I think that how long it is and how we deal with the documentation balances things fairly well.

wilzbach replied:
FWIW I wasn't tackling the two-year deprecation period, but the period of how long the deprecated symbol still stays on dlang.org (I don't like the long deprecation period, but I wasn't proposing to change it).

> Some folks in the past have pushed for never removing deprecated symbols, and both Andrei and Walter have expressed opinions along those lines at one time or another. 

The State of D survey has different results: https://rawgit.com/wilzbach/state-of-d/master/report.html#180373330 (only 2% of the respondents said "never break code")

> After trial and error, we came to the process of having symbols deprecated and documented for one year and then deprecated and undocumented for one year, and over all, that's worked fairly well. 
> ...
> it balances the need for folks to look at the documentation for the symbols they're replacing and the need to discourage the use of such symbols by not having them documented.

Well, the only thing I was questioning whether we can remove the symbol earlier from the docs as two major things changed:
1) docarchives.dlang.io
2) We require visible changelog entries for every deprecation

> and it's a sign that we keep screwing up what we put in Phobos.

Well, two weeks ago I made a "small" list of things that I would change if we can do breaking changes in Phobos. Unfortunately it's a very long list and I stopped at point 135 and being not even through a quarter of Phobos's symbols. 

Anyhow I don't want to block this DIP with this, I just wanted to say that I personally feel that symbols should be undocumented sooner as they are still well visible for people searching for then and removing them helps to avoid encouragement of using deprecated symbols.

jmdavis replied:
Having access to older documentation helps, but if the symbols are undocumented too quickly, it becomes harder for folks updating their code to figure out what the old symbol did exactly so that they can properly update their code. Having them documented for a year and then undocumented for a year came out of trying to balance the desire to remove the documentation as quickly as possible so that folks definitely wouldn't use the symbol and the desire to keep the documentation around the entire time so that folks could easily find it to figure out how to update their code to not use it. There have been folks who have argued in favor of removing the documentation ASAP and others have argued for keeping it around longer. IIRC, at one point, Walter even wanted to have the documentation around _after_ the symbol was gone. The current approach was the compromise between the two extremes.

Now, we didn't use to have a way to access older documentation, so that does arguably change things, but I'm not sure that it's a good idea to speed up removing the documentation, especially since most folks are going to look at the main documentation, and I don't think that very many folks even know about docarchives.dlang.io. Having older versions of the documentation available has been discussed previously, but I think that this is the first that I've heard about there being anywhere that had it other than downloading the zip file for the specific compiler version. How would someone even know to look at docarchives.dlang.io to find older documentation?

ghost91- commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,121 @@
+2. The existance of the current code precludes its own fix or the fix of an equally
+important issue.
+
+Shorting the deprecation period should be done with caution to avoid giving D
```
Shortening

ghost91- commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,121 @@
+
+Both at the time of deprecation and removal, a changelog entry must be made. This
+changelog entry should have a short motivation for the deprecation (or removal)
+and should to describe what steps can be taken by the user to upgrade their codebase.
```
should describe which

mihails-strasuns-sociomantic commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,121 @@
+user code depends on, e.g. [Issue 10378](https://issues.dlang.org/show_bug.cgi?id=10378).
+Fixing such issues should include a deprecation period for the current behavior,
+and an introduction of the new behavior as the default only at the end of the
+period.
```
<3

rjframe commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,121 @@
+deprecation and removal plan. The documentation should contain information to help
+the users using the symbol(s) transition their code away from the symbol(s).
+
+If the deprecation is occuring because the symbol(s) are being replaced by new
```
I don't understand the idea of keeping both old and new code in an un-deprecated state when its known that the old code will be deprecated - what's the reasoning behind this?

It looks like it will just increase the creation of legacy code. For [a trivial] example, if we were to rename std.algorithm.searching's "canFind" to "contains", I would want to know that canFind was deprecated as soon as contains was available so I don't keep writing canFind until the next release, and now have to change even more code. For a mere name substitution, this really doesn't matter, but if I need to do more it would be annoying.

Many people that want their projects to build on multiple compiler versions seem to go for at least three releases, so a single un-deprecated release doesn't solve the problem of supporting multiple versions with the same code; many people will need use version() statements anyway.

CyberShadow replied:
Another case is supporting e.g. both DMD and LDC. It's not reasonably possible without a window.

rjframe replied:
@jmdavis @CyberShadow Would it be too much overhead to document the old code as deprecated, but not mark it in the source until the next version? This way the compiler is happy and its clear to someone reading the docs that they don't want to use the old function.

jmdavis replied:
That's basically how this works now. If a new symbol is replacing an old one, then the new one gets added, and a note gets added to the documentation for the old one that it's going to be deprecated in the next release. Then, in the next release, it's deprecated, and the normal deprecation cycle starts.

If no new symbol is being introduced to replace the symbol being deprecated, then it is deprecated immediately, and the normal deprecation cycle starts.

CyberShadow commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,121 @@
+
+1. The code or feature is notably dangerous or unsafe, and users need to remove
+it from their code as soon as possible.
+2. The existance of the current code precludes its own fix or the fix of an equally
```
Typo, existence

JackStouffer commented:
@CyberShadow The deprecation happens immediately unless it's being replaced with another symbol. In the case of a replacement, then one release window is given for both the new and the old version to exist side by side un-deprecated.

For replacement:

1. Add future replacement and note replacement in docs (one release)
2. Deprecate old symbol (five releases)
3. Removed old symbol's docs; keep deprecation (five releases)
4. Remove.

Without replacement

1. Deprecate old symbol (five releases)
2. Removed old symbol's docs (five releases)
3. Remove.

I wouldn't rely on deprecating in documentation. The chances that people will see that vs a big teal deprecation message in their build log is low.

CyberShadow commented:
> I wouldn't rely on deprecating in documentation. The chances that people will see that vs a big teal deprecation message in their build log is low.

Well, it depends on how you want to balance the goal of forcing people to update their code ASAP vs. allowing the same program to be built with a reasonable number of compiler versions.

The motivation behind a larger document-don't-deprecate window is not whether people will update it or not, but that once they are told do update, their code will continue to build with N compiler versions before then. This includes frontends from other vendors, as well as compilers provided by distribution platforms other than direct dlang.org downloads, such as Linux distro repo packages which can lag behind as well.

CyberShadow commented:
To clarify: once a user starts seeing deprecation message in their compiler output, they now have a dilemma: either update the code immediately, thus immediately breaking it for all compiler versions but N-1, or wait until the symbol is about to be removed, which gives their users a longer window to update their compilers, but causes users of newer compilers, incl. themselves, to suffer through the deprecation messages with each build.

jmdavis commented:
The main problem with delaying deprecation is that relatively few people are going to update their code until they see the deprecation messages. That's how most people are going to realize that they're going to need to update their code. As such, the clock for updating code basically starts ticking once the symbol is actually deprecated regardless of whether a replacement exists before that. And I'd argue that the main reason for the length of the deprecation cycle is because of folks who don't update their code frequently. By having it longer, fewer folks are going to run into the problem of rebuilding their code for the first time in a while and finding that it doesn't compile. By having the symbol deprecated for roughly two years, most folks who actively use D are going to update before the deprecated symbol is removed. How long a replacement symbol is available prior to that period is really orthogonal to that. So, if the period that a replacement symbol has be around before the old one is deprecated is increased, then that should have no effect on how long the symbol is actually deprecated. I would hope though that at this point, replacing one symbol with another would not be a common occurrence, especially since simply renaming a symbol is usually rejected now.

mihails-strasuns-sociomantic commented:
What @CyberShadow suggests has worked pretty well for internal libraries we manage - though we only did it for more annoying changes which easily create hundreds of deprecation messages and more.

> The main problem with delaying deprecation is that relatively few people are going to update their code until they see the deprecation messages.

Yes, this is totally expected. Idea is that by the time library maintainer sees deprecation message, proposed replacement will has already been available for several versions. Which means switching to it is no-brainer even if library wants to support several versions at once. From that perspective best amount of versions to wait before actual deprecation is exactly amount of versions popular libraries support - but not more.

quickfur commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,126 @@
+
+On the first release in the deprecation period, the removed symbol(s) should
+be removed from any module or package wide list of public functions/booktables/cheatsheets
+to demphize its use. On the fifth release in the deprecation period, the documentation
```
Typo: `demphize` -> `deemphasize`.

jacob-carlborg commented:
As far as I can see this DIP doesn’t mention the decision process of how a symbol is deprecated. The assumption here seems to be that it’s already decided that it will be deprecated.

JackStouffer commented:
That's correct. That decision is not part of this DIPs scope.

mdparker commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,121 @@
+## Rationale
+
+There is general disagreement on the best and/or accepted way to remove public
+features. Each deprecation ends up being handled slightly differently depending on
```
"slightly differently" -> with slight variations

mdparker commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,121 @@
+features. Each deprecation ends up being handled slightly differently depending on
+who's handling the pull requests. Standardizing the process makes sure that
+deprecations are done very publicly and carefully, so as to minimize breakage and
+to provide clear fixes for user code.
```
"to provide" -> provide

mdparker commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,121 @@
+
+## Description
+
+The following lays out the rules to follow to deprecate a part of D.
```
"the rules to follow to deprecate" -> a new procedure for deprecating

mdparker commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,121 @@
+
+The following lays out the rules to follow to deprecate a part of D.
+
+A symbol or feature must not be marked for removal on a specific date, but rather on a
```
"on a specific release" -> for a specific release

JackStouffer replied:
Things happen on a date, not for a date.

mdparker commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,121 @@
+The following lays out the rules to follow to deprecate a part of D.
+
+A symbol or feature must not be marked for removal on a specific date, but rather on a
+specific release. This allows users to easily know if upgrading will break their
```
"easily know" -> easily determine

mdparker commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,121 @@
+
+A symbol or feature must not be marked for removal on a specific date, but rather on a
+specific release. This allows users to easily know if upgrading will break their
+code or not.
```
"code or not" -> code.

mdparker commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,121 @@
+an image of instability.
+
+At the time of the pull request for deprecation, all code in Phobos, Druntime,
+and DMD must be updated to remove use of the effected code. Any projects that
```
"remove use of" -> remove all uses of

mdparker replied:
"effected" -> affected

mdparker commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,121 @@
+changelog entry should have a short motivation for the deprecation (or removal)
+and should to describe what steps can be taken by the user to upgrade their codebase.
+
+In order to facilitate on schedule deprecations, a comment of the format
```
"on schedule" -> on-schedule

mdparker commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,121 @@
+and should to describe what steps can be taken by the user to upgrade their codebase.
+
+In order to facilitate on schedule deprecations, a comment of the format
+`@@@DEPRECATED_[version]@@@` should be added to the top of the code to be removed/disabled.
```
"added to the top of the code" -> prefixed to the line or block of code

"to be removed/disabled" -> which is being deprecated.

mdparker commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,121 @@
+
+At the time of the pull request for deprecation, all code in Phobos, Druntime,
+and DMD must be updated to remove use of the effected code. Any projects that
+are tested on the Project Tester that are broken should also have their
```
"Any projects that are tested on the Project Tester that are broken should also have their maintainers notified." ->
Maintainers of any projects which are tested on the Project Tester and are broken by the deprecation should also be notified.

mdparker commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,121 @@
+In order to facilitate on schedule deprecations, a comment of the format
+`@@@DEPRECATED_[version]@@@` should be added to the top of the code to be removed/disabled.
+This comment allows code to be easily searched before every release to
+catch all planned deprecations.
```
"catch" -> find

mdparker commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,121 @@
+
+In order to facilitate on schedule deprecations, a comment of the format
+`@@@DEPRECATED_[version]@@@` should be added to the top of the code to be removed/disabled.
+This comment allows code to be easily searched before every release to
```
"This comment allows code to be easily searched" -> This comment allows for convenient searching within source files

mdparker commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,121 @@
+
+### Public Functions, Types, and Modules
+
+All removals (or changes that make the code `private`/`package`) of public functions,
```
"All removals (or changes that make the code `private`/`package`) of public functions, types, and modules must be accompanied with a deprecation period." -> When public functions, types, and modules are removed, or their protection attributes are made more restrictive, the changes must be accompanied by a deprecation period.

I notice this leaves out variables. Would it be correct to say "public symbols" rather than "functions, types and modules"?

mdparker commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,121 @@
+types, and modules must be accompanied with a deprecation period.
+
+The symbol(s) must be marked using the `deprecated` keyword with a message containing
+the planned removal period. A pointer to more information should also be added. E.g.
```
"pointer" -> reference

mdparker commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,121 @@
+"See the 2.080 changelog for more details" or "See the function documentation for more
+details". The documentation of the symbol(s) must be updated noting the
+deprecation and removal plan. The documentation should contain information to help
+the users using the symbol(s) transition their code away from the symbol(s).
```
"the users using" -> users of

CyberShadow replied:
@mdparker Any reason you're not using one review with multiple comments, and instead posting individual comments? (Everyone here gets one email for each comment above.)

mdparker commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,121 @@
+the users using the symbol(s) transition their code away from the symbol(s).
+
+If the deprecation is occuring because the symbol(s) are being replaced by new
+symbols, both the old and the new symbol(s) should be availible un-deprecated
```
"availible un-deprecated" -> available without deprecation

mdparker commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,121 @@
+
+If there is no equivalent for the functionality of the removed symbol in the
+standard library or the runtime, the code should be moved to
+[undeaD](https://github.com/dlang/undeaD) to allow users to keep their current
```
"to [undeaD] to allow" -> to the undeaD repository, thereby allowing

mdparker commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,121 @@
+Unless the removed language feature is very unsafe or causes damage to real
+world systems, all changes or removals must be accompanied with a deprecation
+period. "Language features" includes bugs in the current behavior that existing
+user code depends on, e.g. [Issue 10378](https://issues.dlang.org/show_bug.cgi?id=10378).
```
"that existing user code depends on" -> upon which existing code depends

The dangling preposition rule in and of itself is nonsense, but in formal writing I do find it distracting.

mdparker commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,121 @@
+period.
+
+Deprecations to language features must also update the [language deprecations
+page](https://dlang.org/deprecate.html) on dlang.org simultaneously. The deprecation
```
Simultaneously to what?

mdparker commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,121 @@
+Deprecations to language features must also update the [language deprecations
+page](https://dlang.org/deprecate.html) on dlang.org simultaneously. The deprecation
+message given by the compiler should contain the planned removal period and/or a
+pointer to more information pertaining to the deprecation.
```
"pointer" -> reference

mdparker commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,121 @@
+
+Warnings must NOT be used in the deprecation process. Warnings are set as errors
+in many build systems (including DUB), and would therefore prematurely break many
+user's code. The exception is when the deprecation is for a change which turns 
```
"many user's code" ->many users' code -> (alternatively): much user code

JackStouffer replied:
This seems like a style preference. "I crashed many people's cars." is a valid sentence.

mdparker commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,121 @@
+Warnings must NOT be used in the deprecation process. Warnings are set as errors
+in many build systems (including DUB), and would therefore prematurely break many
+user's code. The exception is when the deprecation is for a change which turns 
+something into a warning. In this case the code which would trigger the warning must
```
"which turns something into a warning"

I think some clarification is needed here. I had to go back and read it twice. Perhaps "which causes warnings to be emitted where none were before."

mdparker commented:
No reason other than cluelessness. Apologies for the spam!

On Wed, Mar 28, 2018 at 5:07 PM, Vladimir Panteleev <
notifications@github.com> wrote:

> *@CyberShadow* commented on this pull request.
> ------------------------------
>
> In DIPs/DIP1xxx-JS.md
> <https://github.com/dlang/DIPs/pull/108#discussion_r177668137>:
>
> > +In order to facilitate on schedule deprecations, a comment of the format
> +`@@@DEPRECATED_[version]@@@` should be added to the top of the code to be removed/disabled.
> +This comment allows code to be easily searched before every release to
> +catch all planned deprecations.
> +
> +### Public Functions, Types, and Modules
> +
> +All removals (or changes that make the code `private`/`package`) of public functions,
> +types, and modules must be accompanied with a deprecation period.
> +
> +The symbol(s) must be marked using the `deprecated` keyword with a message containing
> +the planned removal period. A pointer to more information should also be added. E.g.
> +"See the 2.080 changelog for more details" or "See the function documentation for more
> +details". The documentation of the symbol(s) must be updated noting the
> +deprecation and removal plan. The documentation should contain information to help
> +the users using the symbol(s) transition their code away from the symbol(s).
>
> @mdparker <https://github.com/mdparker> Any reason you're not using one
> review with multiple comments, and instead posting individual comments?
> (Everyone here gets one email for each comment above.)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dlang/DIPs/pull/108#discussion_r177668137>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA6EDJITr2sNjWpDN19K8ZO6Xije_t70ks5ti0TQgaJpZM4Rz--4>
> .
>



-- 
Mike Parker

wilzbach commented on DIPs/DIP1xxx-JS.md:
```diff
@@ -0,0 +1,126 @@
+"See the 2.080 changelog for more details" or "See the function documentation for more
+details". The documentation of the symbol(s) must be updated noting the
+deprecation and removal plan. The documentation should contain information to help
+the users of the symbol(s) transition their code away from the symbol(s).
```
FYI: I finally got around proposing a `DEPRECATED` ddoc macro for consistent and prettier formatting of of deprecation messages in the spec + making it easier to grep for still documented and deprecated symbols.

https://github.com/dlang/dlang.org/pull/2308

Not sure if it's too late to incorporate this in the DIP.

wilzbach commented:
I just bumped into NodeJS's deprecation page and the majority of the deprecations are "Documentation-only"

https://nodejs.org/api/deprecations.html

So if we ever want to deprecate something, but the cost would be too high or it would bring too little value, just undocumenting it might be an easy way out.
Might be worthwhile to think about this in the DIP.

jmdavis commented:
> So if we ever want to deprecate something, but the cost would be too high or it would bring too little value, just undocumenting it might be an easy way out.

The main problem with that is that you still have to maintain it, so it doesn't help much. It does reduce how much the symbol would be used, but it would still be used, and we'd still have to deal with it. So, IMHO, it doesn't really fix the problem. If it's really bad enough that it needs to be deprecated, then it should be deprecated. And if it's not bad enough for it to be deprecated, then why remove it from the docs? And if it's so heavily used that we don't consider it reasonable to deprecate it, then removing it from the docs is just going to make it harder for folks to figure out how to deal with it when they encounter it, which they will, if it's that heavily used.

wilzbach commented:
> And if it's so heavily used that we don't consider it reasonable to deprecate it, then removing it from the docs is just going to make it harder for folks to figure out how to deal with it when they encounter it, which they will, if it's that heavily used.

The NodeJS "documentation-only" approach is to leave the symbol in the doc, but mark is as deprecated permanently in the docs (that's why I thought that approach is interesting for things that we know are bad, but don't won't to trigger deprecation messages for it). Anyhow, it was just a pointer.

> then why remove it from the docs? 

So that new code that gets written doesn't use this.

mdparker commented:
Alright, I think we've covered the bases for the Draft Review. @JackStouffer If you're ready to go, we can brand this as DIP 1013 and launch the Community Review on Monday.
